### DISCUSSION TITLE
Test not found

### DISCUSSION THREAD

**Discussion by kevgliss** [schemathesis/schemathesis#1232] [Q&A] [answered]
Hi, I'm trying to integrate schemathesis into an existing project and having trouble understanding the following error:

```
ERROR: not found: /home/kglisson/Workspace/dispatch/tests/api/test_schema.py::test_api[DELETE /api/v1/organizations/{organization_id}][P]
(no name '/home/kglisson/Workspace/dispatch/tests/api/test_schema.py::test_api[DELETE /api/v1/organizations/{organization_id}][P]' in any of [<Module tests/api/test_schema.py>]
```

I took the initial setup from the docs:

```
import schemathesis
from dispatch.main import app

schemathesis.fixups.install(["fast_api"])

schema = schemathesis.from_asgi("/api/v1/docs/openapi.json", app, base_url="/api/v1")


@schema.parametrize()
def test_api(case):
    case.call_and_validate()
```

Any hints on what not found here means? Is it a result of the server returning a 404? Or is it something not defined in my test case?

**Comment by Stranger6667** [comment]
Hi @kevgliss 

Thanks for reporting it! The code seems absolutely alright to me. Could you please, share versions of the used Python interpreter, pytest, Schemathesis, and hypothesis?

It is possible that there could be some interference with pytest plugins, so the full output will be helpful as well (it should include all plugins there). Additionally, what command do you use to run the tests?

**Reply by kevgliss** [reply]
Of course, here are the env details:

```
platform linux -- Python 3.8.7, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kglisson/Workspace/dispatch, configfile: setup.cfg
plugins: playwright-0.1.2, schemathesis-3.9.6, Faker-8.8.2, base-url-1.4.2, subtests-0.5.0, hypothesis-6.14.
```

**Reply by kevgliss** [reply]
fwiw I was also able to recreate with what I think are the bare minimum plugin set:

```
patform linux -- Python 3.8.7, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /home/kglisson/Workspace/dispatch, configfile: setup.cfg
plugins: schemathesis-3.9.6, base-url-1.4.2, subtests-0.5.0, hypothesis-6.14.3
```

**Reply by kevgliss** [reply]
One more data point, it seems using the schemathesis test runner performs more like I'd expect:

```
schemathesis run --checks all http://127.0.0.1:8080/api/v1/docs/openapi.json -b "http://127.0.0.1:8080/api/v1"
```

```
===================================================================== SUMMARY =====================================================================

Performed checks:
    not_a_server_error                              173 / 293 passed          FAILED 
    status_code_conformance                         0 / 293 passed            FAILED 
    content_type_conformance                        293 / 293 passed          PASSED 
    response_headers_conformance                    293 / 293 passed          PASSED 
    response_schema_conformance                     293 / 293 passed          PASSED 

========================================================= 12 failed, 102 errored in 41.30s ========================================================
```

**Reply by kevgliss** [reply]
Same error without using ASGI either:

```
import schemathesis

schema = schemathesis.from_uri(
    "http://127.0.0.1:8080/api/v1/docs/openapi.json", base_url="http://127.0.0.1:8080/api/v1"
)

@schema.parametrize()
def test_api(case):
    case.call_and_validate()
```

**Reply by Stranger6667** [reply]
Strange, does it work without the `base-url` plugin?

**Reply by Stranger6667** [reply]
The schema itself is loaded, but then `pytest` fails during collection. Could you share the `pytest` output with the `--collect-only` option?

> (no name '/home/kglisson/Workspace/dispatch/tests/api/test_schema.py::test_api[DELETE /api/v1/organizations/{organization_id}][P]' in any of [<Module tests/api/test_schema.py>]

An additional way to debug it is to check the `pytestdebug.log` file that is created when the `--debug` option is passed to `pytest`.  Let me know if you could share it

**Reply by kevgliss** [reply]
I think it's due to the way VSCode was attempting to run the test, as I've verified that running the command with both pytest and the schemathesis cli don't have the issue.

**Reply by kevgliss** [reply]
This looks relevant: https://github.com/microsoft/vscode-python/issues/3585

**Reply by kevgliss** [reply]
Regardless, this seems like more of vscode issue than a schemathesis issue.

**Comment by kevgliss** [comment, marked as answer]
This was not an issue with schemathesis but with VSCode.

**Reply by Stranger6667** [reply]
Thank you for sharing your findings! :)

**Comment by eduardogallifaochoa** [comment]
Hello,

**TL;DR**  
This wasn’t a Schemathesis bug; it was VS Code’s test runner. VS Code tried to run a *single* parametrized test by passing the full node id (which contains spaces, slashes, and braces) without proper quoting, so `pytest` said “not found” even though the test was collected.

---

## What’s happening
- `@schema.parametrize()` generates ids like  
  `test_api[DELETE /api/v1/organizations/{organization_id}]`.
- The VS Code Python test UI invokes `pytest` with that id literally, and the quoting/escaping breaks → `not found: …`.

---

## How to confirm
- `pytest -q --collect-only` shows the test **is** collected (so it’s not a 404 or a missing test).
- Running from the terminal works:

```bash
pytest -q tests/api/test_schema.py       # run the module
pytest -q -k test_api                    # select the test by name pattern
```

The Schemathesis CLI also works (note the flag change in v4.x):
```bash
schemathesis run http://127.0.0.1:8000/openapi.json \
  --url http://127.0.0.1:8000 \
  --checks all \
  --max-examples 10
```

---
## Workarounds / fixes

- Run tests from a terminal instead of the VS Code test UI, or update the VS Code Python extension (bug tracked in microsoft/vscode-python#3585).
- Avoid selecting a single parametrized case from the UI; use -k filters from the terminal.
- If you’re on Schemathesis v4.x, the CLI uses --url (not --base-url).

Minimal example for v4.x:
```python
import schemathesis

schema = schemathesis.openapi.from_uri("http://127.0.0.1:8000/openapi.json")

@schema.parametrize()
def test_api(case):
    # Provide the base URL here
    case.call_and_validate(base_url="http://127.0.0.1:8000")
```
---
### Windows tip (if you see UnicodeEncodeError in CLI output)

Set UTF-8 for the session before running:
```powershell
$env:PYTHONUTF8="1"
$env:PYTHONIOENCODING="utf-8"
```
```makefile
::contentReference[oaicite:0]{index=0}
```
